test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -89,110 +106,162 @@ final case class Spec[+L, +T](caseValue: SpecCase[L, T, Spec[L, T]]) { self =>
    * effectfully transforming every test with the provided function, finally
    * reconstructing the spec with the same structure.
    */
-  final def foreachPar[R, E, A](f: T => ZIO[R, E, A]): ZIO[R, E, Spec[L, A]] =
-    foreachExec(ExecutionStrategy.Parallel)(f)
+  final def foreachPar[R1 <: R, E1, A](
+    failure: E => ZIO[R1, E1, A],
+    success: T => ZIO[R1, E1, A]
+  ): ZIO[R1, Nothing, Spec[R1, E1, L, A]] =
+    foreachExec(ExecutionStrategy.Parallel)(failure, success)
 
   /**
    * Iterates over the spec with the parallel (`n`) strategy as the default, and
    * effectfully transforming every test with the provided function, finally
    * reconstructing the spec with the same structure.
    */
-  final def foreachParN[R, E, A](f: T => ZIO[R, E, A]): ZIO[R, E, Spec[L, A]] =
-    foreachExec(ExecutionStrategy.Parallel)(f)
+  final def foreachParN[R1 <: R, E1, A](
+    n: Int
+  )(failure: E => ZIO[R1, E1, A], success: T => ZIO[R1, E1, A]): ZIO[R1, Nothing, Spec[R1, E1, L, A]] =
+    foreachExec(ExecutionStrategy.ParallelN(n))(failure, success)
 
   /**
    * Returns a new spec with remapped labels.
    */
-  final def mapLabel[L1](f: L => L1): Spec[L1, T] =
-    transform[L1, T] {
+  final def mapLabel[L1](f: L => L1): Spec[R, E, L1, T] =
+    transform[R, E, L1, T] {
       case SuiteCase(label, specs, exec) => SuiteCase(f(label), specs, exec)
       case TestCase(label, test)         => TestCase(f(label), test)
     }
 
   /**
    * Returns a new spec with remapped tests.
    */
-  final def mapTest[T1](f: T => T1): Spec[L, T1] =
-    transform[L, T1] {
+  final def mapTest[T1](f: T => T1): Spec[R, E, L, T1] =
+    transform[R, E, L, T1] {
       case SuiteCase(label, specs, exec) => SuiteCase(label, specs, exec)
-      case TestCase(label, test)         => TestCase(label, f(test))
+      case TestCase(label, test)         => TestCase(label, test.map(f))
     }
 
   /**
-   * Computes the size of the spec, i.e. the number of tests in the spec.
+   * Uses the specified `Managed` to provide each test in this spec with its
+   * required environment.
    */
-  final def size: Int = fold[Int] {
-    case SuiteCase(_, counts, _) => counts.sum
-    case TestCase(_, _)          => 1
+  final def provideManaged[E1 >: E](managed: Managed[E1, R]): Spec[Any, E1, L, T] =
+    transform[Any, E1, L, T] {
+      case SuiteCase(label, specs, exec) => SuiteCase(label, specs.provideManaged(managed), exec)
+      case TestCase(label, test)         => TestCase(label, test.provideManaged(managed))
+    }
+
+  /**
+   * Uses the specified `Managed` once to provide all tests in this spec with
+   * a shared version of their required environment. This is useful when the
+   * act of creating the environment is expensive and should only be performed
+   * once.
+   */
+  final def provideManagedSuite[E1 >: E](managed: Managed[E1, R]): Spec[Any, E1, L, T] = {
+    def loop(r: R)(spec: Spec[R, E, L, T]): ZIO[Any, E, Spec[Any, E, L, T]] =
```

Comment:
I feel like this `loop` suggests the need for another traversal scheme.

Reply:
I agree. I've been struggling a bit to find the right encoding. I think what we want is a  function that takes a `ZIO[R, E, T]` in the test case to a `ZIO[R1, E1, T]`, where there are no constrains on `R1` and `E1` being subtypes / supertypes of `R` and `E`. And then I think for the suite case we basically need a universally quantified function that takes a `ZIO[R, E, X]` to a `ZIO[R1, E1, X]`, so the suite case can change the effect but should delegate changing the child specs to recursively calling the traversal. Otherwise we run into issues with the environment and error types getting widened to the union of `R` and `R1` and `E` and `E1` instead of just getting `R1 ` and `E`.